File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/between/HouseholdEventListener.kt
package org.wyrdsekai.app.engine.between

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Subscribes to household-wide events on `between.{householdId}.events`.
 *
 * These are low-volume, significant events:
 * - New agent arrived in the household
 * - Agent departed or went dormant
 * - Steward announcements
 * - Safety alerts
 * - Configuration changes
 *
 */
@Serializable
sealed class HouseholdEvent {
    abstract val timestamp: Long

    @Serializable
    @SerialName("agent_arrived")
    data class AgentArrived(
        val agentDid: String,
        val agentName: String,
        override val timestamp: Long,
    ) : HouseholdEvent()

    @Serializable
    @SerialName("agent_departed")
    data class AgentDeparted(
        val agentDid: String,
        val agentName: String,
        val reason: String? = null,
        override val timestamp: Long,
    ) : HouseholdEvent()

    @Serializable
    @SerialName("steward_announcement")
    data class StewardAnnouncement(
        val stewardDid: String,
        val message: String,
        override val timestamp: Long,
    ) : HouseholdEvent()

    @Serializable
    @SerialName("safety_alert")
    data class SafetyAlert(
        val severity: String,
        val message: String,
        val sourceDid: String? = null,
        override val timestamp: Long,
    ) : HouseholdEvent()

    @Serializable
    @SerialName("config_changed")
    data class ConfigChanged(
        val key: String,
        val oldValue: String? = null,
        val newValue: String? = null,
        override val timestamp: Long,
    ) : HouseholdEvent()
}

class HouseholdEventListener(
    private val between: BetweenClient,
    private val householdId: String,
    private val onEvent: (HouseholdEvent) -> Unit,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private var unsubscribe: (() -> Unit)? = null

    /** Start listening for household events. */
    fun startListening() {
        val subject = "between.$householdId.events"
        unsubscribe = between.subscribe(subject) { _, data ->
            try {
                val event = json.decodeFromString<HouseholdEvent>(data.decodeToString())
                onEvent(event)
            } catch (_: Exception) {
                // Malformed household event — skip
            }
        }
    }

    /** Stop listening for household events. */
    fun stopListening() {
        unsubscribe?.invoke()
        unsubscribe = null
    }
}

File: clients/rn/src/engine/between/HouseholdEventListener.ts
/**
 * Subscribes to household-wide events on `between.{householdId}.events`.
 *
 * These are low-volume, significant events:
 * - New agent arrived in the household
 * - Agent departed or went dormant
 * - Steward announcements
 * - Safety alerts
 * - Configuration changes
 *
 */

import type { BetweenClient } from './BetweenClient';

export type HouseholdEvent =
  | AgentArrived
  | AgentDeparted
  | StewardAnnouncement
  | SafetyAlert
  | ConfigChanged;

export interface AgentArrived {
  type: 'agent_arrived';
  agentDid: string;
  agentName: string;
  timestamp: number;
}

export interface AgentDeparted {
  type: 'agent_departed';
  agentDid: string;
  agentName: string;
  reason?: string;
  timestamp: number;
}

export interface StewardAnnouncement {
  type: 'steward_announcement';
  stewardDid: string;
  message: string;
  timestamp: number;
}

export interface SafetyAlert {
  type: 'safety_alert';
  severity: string;
  message: string;
  sourceDid?: string;
  timestamp: number;
}

export interface ConfigChanged {
  type: 'config_changed';
  key: string;
  oldValue?: string;
  newValue?: string;
  timestamp: number;
}

export type HouseholdEventCallback = (event: HouseholdEvent) => void;

export class HouseholdEventListener {
  private unsubscribe: (() => void) | null = null;

  constructor(
    private readonly between: BetweenClient,
    private readonly householdId: string,
    private readonly onEvent: HouseholdEventCallback,
  ) {}

  /** Start listening for household events. */
  startListening(): void {
    const subject = `between.${this.householdId}.events`;
    this.unsubscribe = this.between.subscribe(subject, (_subject, data) => {
      try {
        const event = JSON.parse(new TextDecoder().decode(data)) as HouseholdEvent;
        this.onEvent(event);
      } catch {
        // Malformed household event — skip
      }
    });
  }

  /** Stop listening for household events. */
  stopListening(): void {
    this.unsubscribe?.();
    this.unsubscribe = null;
  }
}
